server/src/main/java/io/crate/role/TransportPrivilegesAction.java

```diff
@@ -91,61 +92,76 @@ protected void masterOperation(PrivilegesRequest request,
+        if (request.rolePrivilege() != null) {
+            roleNamesToCheck.addAll(request.rolePrivilege().roleNames());
+        }
         if (rolesMetadata == null) {
-            return new ArrayList<>(userNames);
+            return new ArrayList<>(roleNamesToCheck);
         }
-        List<String> unknownUserNames = null;
-        for (String userName : userNames) {
+        List<String> unknownRoleNames = null;
+        for (String roleName : roleNamesToCheck) {
             //noinspection PointlessBooleanExpression
-            if (rolesMetadata.roleNames().contains(userName) == false) {
-                if (unknownUserNames == null) {
-                    unknownUserNames = new ArrayList<>();
+            if (rolesMetadata.roleNames().contains(roleName) == false) {
+                if (unknownRoleNames == null) {
+                    unknownRoleNames = new ArrayList<>();
                 }
-                unknownUserNames.add(userName);
+                unknownRoleNames.add(roleName);
             }
         }
-        if (unknownUserNames == null) {
+        if (unknownRoleNames == null) {
             return Collections.emptyList();
         }
-        return unknownUserNames;
+        return unknownRoleNames;
     }
 
     @VisibleForTesting
     static long applyPrivileges(Metadata.Builder mdBuilder,
                                 PrivilegesRequest request) {
-        // create a new instance of the metadata, to guarantee the cluster changed action.
-        UsersPrivilegesMetadata newMetadata = UsersPrivilegesMetadata.copyOf(
-            (UsersPrivilegesMetadata) mdBuilder.getCustom(UsersPrivilegesMetadata.TYPE));
-
-        long affectedRows = newMetadata.applyPrivileges(request.userNames(), request.privileges());
-        mdBuilder.putCustom(UsersPrivilegesMetadata.TYPE, newMetadata);
-        return affectedRows;
+        long affectedRows;
+        if (request.privileges().isEmpty() == false) {
+            // create a new instance of the metadata, to guarantee the cluster changed action.
+            UsersPrivilegesMetadata newMetadata = UsersPrivilegesMetadata.copyOf(
+                (UsersPrivilegesMetadata) mdBuilder.getCustom(UsersPrivilegesMetadata.TYPE));
+
+            affectedRows = newMetadata.applyPrivileges(request.userNames(), request.privileges());
+            mdBuilder.putCustom(UsersPrivilegesMetadata.TYPE, newMetadata);
+            return affectedRows;
+        } else {
+            // create a new instance of the metadata, to guarantee the cluster changed action.
+            RolesMetadata newMetadata = RolesMetadata.newInstance(
+                (RolesMetadata) mdBuilder.getCustom(UsersPrivilegesMetadata.TYPE));
+
+            affectedRows = newMetadata.applyPrivileges(request.userNames(), request.rolePrivilege());
+            mdBuilder.putCustom(UsersPrivilegesMetadata.TYPE, newMetadata);
+            return affectedRows;
+        }
     }
```

Comment:
in if/else - wanted to add a comment that we need to check that privileges/roles are mutually exclusive, but then found we do it already in `AnalyzedPrivileges`
> "privileges and rolePrivileges cannot be set together"

However, I didn't see a test for that. Could you please add one if it's missing? Sorry if it exists already but I couldn't find.

Reply:
https://github.com/crate/crate/pull/15249/files#diff-7a4fe3442cba906f593d1cbe00180e547da987aa37207cdb51b697ef94e8e5bdR344